crates/bevy_asset/src/asset_changed.rs

```diff
@@ -0,0 +1,475 @@
+//! Filter query to limit iteration over [`Handle`]s which content was updated recently.
+use std::marker::PhantomData;
+
+use bevy_ecs::{
+    archetype::{Archetype, ArchetypeComponentId},
+    component::{ComponentId, Tick},
+    prelude::{Changed, Entity, Or, Resource, World},
+    query::{Access, FilteredAccess, QueryItem, ReadFetch, ReadOnlyWorldQuery, WorldQuery},
+    storage::{Table, TableRow},
+    world::unsafe_world_cell::UnsafeWorldCell,
+};
+use bevy_utils::{get_short_name, HashMap};
+
+use crate::{Asset, AssetId, Handle};
+
+#[doc(hidden)]
+#[derive(Resource)]
+pub struct AssetChanges<A: Asset> {
+    change_ticks: HashMap<AssetId<A>, Tick>,
+    last_change_tick: Tick,
+}
+
+impl<A: Asset> AssetChanges<A> {
+    pub(crate) fn insert(&mut self, asset_id: AssetId<A>, tick: Tick) {
+        self.last_change_tick = tick;
+        self.change_ticks.insert(asset_id, tick);
+    }
+    pub(crate) fn remove(&mut self, asset_id: &AssetId<A>) {
+        self.change_ticks.remove(asset_id);
+    }
+}
+impl<A: Asset> Default for AssetChanges<A> {
+    fn default() -> Self {
+        Self {
+            change_ticks: Default::default(),
+            last_change_tick: Tick::new(0),
+        }
+    }
+}
+
+struct AssetChangeCheck<'w, A: Asset> {
+    handle_change_map: &'w HashMap<AssetId<A>, Tick>,
+    last_run: Tick,
+    this_run: Tick,
+}
+impl<A: Asset> Clone for AssetChangeCheck<'_, A> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+impl<A: Asset> Copy for AssetChangeCheck<'_, A> {}
+impl<'w, A: Asset> AssetChangeCheck<'w, A> {
+    fn new(changes: &'w AssetChanges<A>, last_run: Tick, this_run: Tick) -> Self {
+        Self {
+            handle_change_map: &changes.change_ticks,
+            last_run,
+            this_run,
+        }
+    }
+    // TODO(perf): some sort of caching? Each check has two levels of indirection,
+    // which is not optimal.
```

Comment:
Maybe use `AssetChanges::last_change_tick` (which I understand is common to all asset instances of type `A`) as an early out test before hitting the hash map for the particular asset instance? If `A` is a type of asset that doesn't change often, this will completely avoid the hash map lookup most of the time.

Reply:
Yep, that's a good idea. We are currently doing this, it's in the `WorldQuery::init_fetch` impl of `AssetChanged`